File: day-4/ts/part-2.ts
"use strict";
import fs from "node:fs";
import { checkItem } from "./part-1.ts";

export function totalRollsRemoved(filePath = "../input.txt") {
  const fileContent: string = fs.readFileSync(filePath, "utf-8");
  const lines = fileContent.split("\n");

  const grid: string[][] = [];
  lines.forEach((line) => grid.push(line.split("")));

  let hasRollsToRemove = true;
  let removedRollsPositions: string[] = [];
  let totalRolls = 0;

  while (hasRollsToRemove) {
    let rolls = 0;

    for (let r = 0; r < grid.length; r++) {
      for (let c = 0; c < grid[r]!.length; c++) {
        if (grid[r]![c] !== "@") continue;

        let adjacentItems = 0;
        if (checkItem(grid, r - 1, c - 1)) adjacentItems += 1;
        if (checkItem(grid, r - 1, c)) adjacentItems += 1;
        if (checkItem(grid, r - 1, c + 1)) adjacentItems += 1;
        if (checkItem(grid, r, c - 1)) adjacentItems += 1;
        if (checkItem(grid, r, c + 1)) adjacentItems += 1;
        if (checkItem(grid, r + 1, c - 1)) adjacentItems += 1;
        if (checkItem(grid, r + 1, c)) adjacentItems += 1;
        if (checkItem(grid, r + 1, c + 1)) adjacentItems += 1;

        if (adjacentItems < 4) {
          const pos = r + "," + c;
          removedRollsPositions.push(pos);
          rolls += 1;
        }
      }
    }

    if (rolls === 0) hasRollsToRemove = false;
    totalRolls += rolls;
    removedRollsPositions.forEach((position) => {
      const [r, c] = position.split(",").map(Number);
      grid[r][c] = ".";
    });
    removedRollsPositions = [];
  }

  return totalRolls;
}

totalRollsRemoved();

File: day-4/go/part2.go
package main

func solvePart2(grid [][]string) int {
	count := 0

	for {
		removals := make([][2]int, 0)

		for r := 0; r < len(grid); r++ {
			for c := 0; c < len(grid[r]); c++ {
				if grid[r][c] != "@" {
					continue
				}

				adjacentItems := 0
				for _, pos := range adjacentItemPositions {
					if checkItem(grid, r+pos[0], c+pos[1]) {
						adjacentItems += 1
					}
				}
				if adjacentItems < 4 {
					removals = append(removals, [2]int{r, c})
				}
			}
		}

		if len(removals) == 0 {
			break
		}
		count += len(removals)
		for _, p := range removals {
			grid[p[0]][p[1]] = "."
		}
	}

	return count
}
